### PR TITLE
Track Jets Emulator and Track Jet Word Format

### PR DESCRIPTION
**Pull request by samer973**
**PR description:**

This PR does the following:

1.    Adds in a TkJetWord data format to be used as the output to the track jets emulation.
2.    Adds an emulation for the track jets firmware. This emulation can be run alongside the simulation.
3.    Updates the L1TrackObjectNtupleMaker to store the emulation information.

**Future updates not included in this PR:**

1. In the future we will want the ability to write out text files which contain the emulation inputs and outputs, for comparison to the HLS/firmware code. However, this feature is not yet implemented.
2. We currently take the global phi of each track directly from TTTrack_TrackWord. Once GTT increases the precision of the local phi in TTTrack_TrackWord, we will update the emulator to calculate the global phi of each track from this local phi and the phi sector of each track.

**PR validation:**

This PR has been validated in the following ways:

1.    scram b code-checks
2.    We also created some validation plots to see that the code was behaving as expected.

### CODE REVIEW

File: L1Trigger/L1TTrackMatch/python/L1TrackJetEmulationProducer_cfi.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	nStubs5DisplacedChi2_Loose=cms.double(2.75),
+	nStubs5Displacedbend_Loose=cms.double(3.5),
+	nStubs4DisplacedChi2_Tight=cms.double(12.0),
+	nStubs4Displacedbend_Tight=cms.double(1.0),
+	nStubs5DisplacedChi2_Tight=cms.double(2.75),
+	nStubs5Displacedbend_Tight=cms.double(3.5)
+)
+
+L1TrackJetsExtendedEmulation = cms.EDProducer('L1TrackJetEmulationProducer',
+	L1TrackInputTag= cms.InputTag("L1GTTInputProducerExtended", "Level1TTTracksConverted"),
```
Comment (gkaratha) [inline L1Trigger/L1TTrackMatch/python/L1TrackJetEmulationProducer_cfi.py]: here it should be "Level1TTTracksExtendedConverted" not "Level1TTTracksConverted"

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc [hunk 2]
```diff
@@ -677,6 +709,13 @@ void L1TrackObjectNtupleMaker::beginJob() {
   m_2ltrkjetExt_nTightDisplaced=new std::vector<int>;
   m_2ltrkjetExt_ntdtrk=new std::vector<int>;
 
+  m_trkjetemExt_pt = new std::vector<float>;	
+  m_trkjetemExt_phi = new std::vector<float>;	
+  m_trkjetemExt_eta = new std::vector<float>;	
+  m_trkjetemExt_z = new std::vector<float>;
```
Comment (gkaratha) [inline L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc:715]: Seems that you introduce this variable but then never write the branch, Is it just a negligence?

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py [hunk 3]
```diff
@@ -37,10 +37,10 @@
 # input and output
 ############################################################
 
-process.maxEvents = cms.untracked.PSet(input = cms.untracked.int32(100))
+process.maxEvents = cms.untracked.PSet(input = cms.untracked.int32(-1))
 
 readFiles = cms.untracked.vstring(
-    "/store/relval/CMSSW_11_1_0_pre2/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU25ns_110X_mcRun4_realistic_v2_2026D49PU200-v1/20000/F7BF4AED-51F1-9D47-B86D-6C3DDA134AB9.root"
+    'file:/tmp/TT_TuneCP5_14TeV-powheg-pythia8_Phase2HLTTDRSummer20ReRECOMiniAOD-PU200_111X_mcRun4_realistic_T15_v1-v2_FEVT_015FB6F1-59B4-304C-B540-2392A983A97D.root'
```
Comment (gkaratha) [inline L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py]: Probably better to re-introduce a commonly accessible file so it runs straight out-of-the-box

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py [hunk 4]
```diff
@@ -150,11 +158,14 @@
...
         SaveTrackJets = cms.bool(True),
+	SaveTrackJetsEmulation = cms.bool(True),
         SaveTrackMET = cms.bool(True),
         TrackFastJetsInputTag = cms.InputTag("L1TrackFastJets","L1TrackFastJets"),
         TrackFastJetsExtendedInputTag = cms.InputTag("L1TrackFastJetsExtended","L1TrackFastJetsExtended"),
-        TrackJetsInputTag=cms.InputTag("L1TrackJets", "L1TrackJets"),
-        TrackJetsExtendedInputTag=cms.InputTag("L1TrackJetsExtended", "L1TrackJetsExtended"),
+        TrackJetsInputTag = cms.InputTag("L1TrackJets", "L1TrackJets"),
+	TrackJetsExtendedInputTag=cms.InputTag("L1TrackJetsExtended", "L1TrackJetsExtended"),
+	TrackJetsEmulationInputTag = cms.InputTag("L1TrackJetsEmulation","L1TrackJets"),
```
Comment (gkaratha) [inline L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py:167]: indentation please

File: DataFormats/L1Trigger/interface/TkJetWord.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    int intNTracks() const { return Scales::intNt(hwNt); }
+    int intNXTracks() const { return Scales::intXt(hwXt); }
+    float floatPt() const { return Scales::floatPt(hwPt); }
+    float floatEta() const { return Scales::floatEta(hwEta); }
+    float floatPhi() const { return Scales::floatPhi(hwPhi); }
+
+  };
+ 
+
+  typedef std::vector<l1t::TkJetWord> TkJetEmulationCollection;
```
Comment (gkaratha) [inline DataFormats/L1Trigger/interface/TkJetWord.h]: Why TkJetEmulationCollection and not TkJetWordCollection? I thought the common approach is std::vector<object> == objectCollection, no?

File: DataFormats/L1Trigger/interface/TkJetWord.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    int intPt() const { return Scales::intPt(hwPt); }
+    int intNTracks() const { return Scales::intNt(hwNt); }
+    int intNXTracks() const { return Scales::intXt(hwXt); }
+    float floatPt() const { return Scales::floatPt(hwPt); }
+    float floatEta() const { return Scales::floatEta(hwEta); }
+    float floatPhi() const { return Scales::floatPhi(hwPhi); }
+
+  };
+ 
+
```
Comment (gkaratha) [inline DataFormats/L1Trigger/interface/TkJetWord.h:93]: Also would it make sense to add a ref to the used tracks as in DataFormats/L1TCorrelator/interface/TkJet.h?

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py (hunk 3 above)
Comment (samer973) [reply L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py]: Changed back to commonly accessible file.

File: L1Trigger/L1TTrackMatch/python/L1TrackJetEmulationProducer_cfi.py (hunk 1 above)
Comment (samer973) [reply L1Trigger/L1TTrackMatch/python/L1TrackJetEmulationProducer_cfi.py]: Fixed.

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc (hunk 2 above)
Comment (samer973) [reply L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc:715]: The branch is now written so we can study the z position of our jets.

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py (hunk 4 above)
Comment (samer973) [reply L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py:167]: In the file the indentation is consistent. I'm not sure why this preview has it being misaligned.

File: DataFormats/L1Trigger/interface/TkJetWord.h (hunk 5 above)
Comment (samer973) [reply DataFormats/L1Trigger/interface/TkJetWord.h]: I've changed the name to be in line with this convention. It's now called TkJetWordCollection.

**Pull request closed without merging**